Overlapping [From,To]JSON instances with aeson-1.5.3.0

`hnix` currently fails to compile with `aeson-1.5.3.0`:

<details>

```
src/Nix/Expr/Types.hs:546:10: error:
    • Overlapping instances for ToJSON NExpr
        arising from a use of ‘aeson-1.5.3.0:Data.Aeson.Types.ToJSON.$dmtoJSON’
      Matching instances:
        instance ToJSON1 f => ToJSON (Fix f)
          -- Defined in ‘aeson-1.5.3.0:Data.Aeson.Types.ToJSON’
        instance ToJSON NExpr -- Defined at src/Nix/Expr/Types.hs:546:10
    • In the expression:
        aeson-1.5.3.0:Data.Aeson.Types.ToJSON.$dmtoJSON @(NExpr)
      In an equation for ‘toJSON’:
          toJSON = aeson-1.5.3.0:Data.Aeson.Types.ToJSON.$dmtoJSON @(NExpr)
      In the instance declaration for ‘ToJSON NExpr’
    |
546 | instance ToJSON NExpr
    |          ^^^^^^^^^^^^

src/Nix/Expr/Types.hs:546:10: error:
    • Overlapping instances for ToJSON NExpr
        arising from a use of ‘aeson-1.5.3.0:Data.Aeson.Types.ToJSON.$dmtoEncoding’
      Matching instances:
        instance ToJSON1 f => ToJSON (Fix f)
          -- Defined in ‘aeson-1.5.3.0:Data.Aeson.Types.ToJSON’
        instance ToJSON NExpr -- Defined at src/Nix/Expr/Types.hs:546:10
    • In the expression:
        aeson-1.5.3.0:Data.Aeson.Types.ToJSON.$dmtoEncoding @(NExpr)
      In an equation for ‘toEncoding’:
          toEncoding
            = aeson-1.5.3.0:Data.Aeson.Types.ToJSON.$dmtoEncoding @(NExpr)
      In the instance declaration for ‘ToJSON NExpr’
    |
546 | instance ToJSON NExpr
    |          ^^^^^^^^^^^^

src/Nix/Expr/Types.hs:546:10: error:
    • Overlapping instances for ToJSON NExpr
        arising from a use of ‘aeson-1.5.3.0:Data.Aeson.Types.ToJSON.$dmtoJSONList’
      Matching instances:
        instance ToJSON1 f => ToJSON (Fix f)
          -- Defined in ‘aeson-1.5.3.0:Data.Aeson.Types.ToJSON’
        instance ToJSON NExpr -- Defined at src/Nix/Expr/Types.hs:546:10
    • In the expression:
        aeson-1.5.3.0:Data.Aeson.Types.ToJSON.$dmtoJSONList @(NExpr)
      In an equation for ‘toJSONList’:
          toJSONList
            = aeson-1.5.3.0:Data.Aeson.Types.ToJSON.$dmtoJSONList @(NExpr)
      In the instance declaration for ‘ToJSON NExpr’
    |
546 | instance ToJSON NExpr
    |          ^^^^^^^^^^^^

src/Nix/Expr/Types.hs:546:10: error:
    • Overlapping instances for ToJSON NExpr
        arising from a use of ‘aeson-1.5.3.0:Data.Aeson.Types.ToJSON.$dmtoEncodingList’
      Matching instances:
        instance ToJSON1 f => ToJSON (Fix f)
          -- Defined in ‘aeson-1.5.3.0:Data.Aeson.Types.ToJSON’
        instance ToJSON NExpr -- Defined at src/Nix/Expr/Types.hs:546:10
    • In the expression:
        aeson-1.5.3.0:Data.Aeson.Types.ToJSON.$dmtoEncodingList @(NExpr)
      In an equation for ‘toEncodingList’:
          toEncodingList
            = aeson-1.5.3.0:Data.Aeson.Types.ToJSON.$dmtoEncodingList @(NExpr)
      In the instance declaration for ‘ToJSON NExpr’
    |
546 | instance ToJSON NExpr
    |          ^^^^^^^^^^^^

src/Nix/Expr/Types.hs:561:10: error:
    • Overlapping instances for FromJSON NExpr
        arising from a use of ‘aeson-1.5.3.0:Data.Aeson.Types.FromJSON.$dmparseJSON’
      Matching instances:
        instance FromJSON1 f => FromJSON (Fix f)
          -- Defined in ‘aeson-1.5.3.0:Data.Aeson.Types.FromJSON’
        instance FromJSON NExpr -- Defined at src/Nix/Expr/Types.hs:561:10
    • In the expression:
        aeson-1.5.3.0:Data.Aeson.Types.FromJSON.$dmparseJSON @(NExpr)
      In an equation for ‘parseJSON’:
          parseJSON
            = aeson-1.5.3.0:Data.Aeson.Types.FromJSON.$dmparseJSON @(NExpr)
      In the instance declaration for ‘FromJSON NExpr’
    |
561 | instance FromJSON NExpr
    |          ^^^^^^^^^^^^^^

src/Nix/Expr/Types.hs:561:10: error:
    • Overlapping instances for FromJSON NExpr
        arising from a use of ‘aeson-1.5.3.0:Data.Aeson.Types.FromJSON.$dmparseJSONList’
      Matching instances:
        instance FromJSON1 f => FromJSON (Fix f)
          -- Defined in ‘aeson-1.5.3.0:Data.Aeson.Types.FromJSON’
        instance FromJSON NExpr -- Defined at src/Nix/Expr/Types.hs:561:10
    • In the expression:
        aeson-1.5.3.0:Data.Aeson.Types.FromJSON.$dmparseJSONList @(NExpr)
      In an equation for ‘parseJSONList’:
          parseJSONList
            = aeson-1.5.3.0:Data.Aeson.Types.FromJSON.$dmparseJSONList @(NExpr)
      In the instance declaration for ‘FromJSON NExpr’
    |
561 | instance FromJSON NExpr
    |          ^^^^^^^^^^^^^^
```
</details>

The problem is that our instances for `NExpr` overlap with the new instances for `Fix` [added in `aeson-1.5.3.0`](http://hackage.haskell.org/package/aeson-1.5.3.0/changelog)

https://github.com/haskell-nix/hnix/blob/1a2d627a5807cde657ab1277234e9521008e8114/src/Nix/Expr/Types.hs#L533-L561

While looking into this I noticed that the current instances (added in https://github.com/haskell-nix/hnix/commit/146780c024adb3c9ec3d2122863a90e5fa5d276a) behave somewhat weirdly:

```
> :set -XOverloadedStrings 
> toJSON ("bla" :: NExpr)
Object (fromList [("unFix",Object (fromList [("tag",String "NSym"),("contents",String "bla")]))])
```

Is the `unFix` key an intended part of the output, and should we try preserving it?

Also, what's the purpose of the instances for `NExpr`? What are they being used for?

In any case I haven't figured out yet how to make use of the new `instance ToJSON1 f => ToJSON (Fix f)`. With the following patch, I get a compile failure:

```diff
--- a/src/Nix/Expr/Types.hs
+++ b/src/Nix/Expr/Types.hs
@@ -508,10 +508,11 @@ $(deriveShow1 ''Binding)
 $(deriveShow1 ''Antiquoted)
 $(deriveShow2 ''Antiquoted)
 
---x $(deriveJSON1 defaultOptions ''NExprF)
+$(deriveJSON1 defaultOptions ''NExprF)
+$(deriveJSON1 defaultOptions ''NKeyName)
 $(deriveJSON1 defaultOptions ''NString)
 $(deriveJSON1 defaultOptions ''Params)
---x $(deriveJSON1 defaultOptions ''Binding)
+$(deriveJSON1 defaultOptions ''Binding)
 $(deriveJSON1 defaultOptions ''Antiquoted)
 $(deriveJSON2 defaultOptions ''Antiquoted)
 
@@ -543,7 +544,9 @@ instance ToJSON NUnaryOp
 instance ToJSON NBinaryOp
 instance ToJSON NRecordType
 instance ToJSON a => ToJSON (NExprF a)
+#if !MIN_VERSION_aeson(1,5,3)
 instance ToJSON NExpr
+#endif
 
 instance (FromJSON v, FromJSON a) => FromJSON (Antiquoted v a)
 instance FromJSON a => FromJSON (NString a)
@@ -558,7 +561,9 @@ instance FromJSON NUnaryOp
 instance FromJSON NBinaryOp
 instance FromJSON NRecordType
 instance FromJSON a => FromJSON (NExprF a)
+#if !MIN_VERSION_aeson(1,5,3)
 instance FromJSON NExpr
+#endif
 
 $(makeTraversals ''NExprF)
 $(makeTraversals ''Binding)
```

```
src/Nix/Expr/Types.hs:1:1: error:
    Exception when trying to run compile-time code:
      Constructor ‘DynamicKey‘ must only use its last 1 type variable(s) within the last 1 argument(s) of a data type
```

@jwiegley Any thoughts / advice?

> I'm not sure anyone is even using the JSON instances for `NExpr`, so dropping them should be fine.

Actually, the `ToJSON NExpr` instance is used by the `--json` option:

```
main/Main.hs:127:30: error:
    • No instance for (aeson-1.5.3.0:Data.Aeson.Types.ToJSON.ToJSON1
                         NExprF)
        arising from a use of ‘A.encodeToLazyText’
    • In the second argument of ‘($)’, namely
        ‘A.encodeToLazyText (stripAnnotation expr)’
      In the second argument of ‘($)’, namely
        ‘TL.putStrLn $ A.encodeToLazyText (stripAnnotation expr)’
      In the expression:
        liftIO $ TL.putStrLn $ A.encodeToLazyText (stripAnnotation expr)
    |
127 |     = liftIO $ TL.putStrLn $ A.encodeToLazyText (stripAnnotation expr)
    |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

```
$ hnix --json --expr '-1'
{"unFix":{"tag":"NUnary","contents":["NNeg",{"unFix":{"tag":"NConstant","contents":{"tag":"NInt","contents":1}}}]}}
```

This functionality was added in https://github.com/haskell-nix/hnix/commit/8d25aa461dbfe267292d91c289375a7702e561b9.

The usefulness of this output is still not very obvious to me. I guess it could be used for debugging, but that functionality could be superseded by https://github.com/haskell-nix/hnix/issues/661.

Is it OK if we remove this codepath?

https://github.com/haskell-nix/hnix/blob/0449dc457018b25335046d42be84ca3460feec16/main/Main.hs#L126-L127

The combination of `--json` and `--eval` should of course be preserved – it looks completely reasonable to me:

```
$ hnix --eval --json --expr '1 + 1'
2
```

--- a/hnix.cabal
+++ b/hnix.cabal
@@ -384,7 +384,7 @@ library
       src
   ghc-options: -Wall -fprint-potential-instances
   build-depends:
-      aeson >= 1.4.2 && < 1.5.3
+      aeson >= 1.4.2 && < 1.6
     , array >= 0.4 && < 0.6
     , base >= 4.11 && < 5
     , binary >= 0.8.5 && < 0.9

--- a/main/Main.hs
+++ b/main/Main.hs
@@ -17,7 +17,6 @@ import           Control.Monad.Catch
 import           Control.Monad.Free
 import           Control.Monad.IO.Class
 -- import           Control.Monad.ST
-import qualified Data.Aeson.Text               as A
 import qualified Data.HashMap.Lazy             as M
 import qualified Data.Map                      as Map
 import           Data.List                      ( sortOn )
@@ -25,7 +24,6 @@ import           Data.Maybe                     ( fromJust )
 import           Data.Time
 import qualified Data.Text                     as Text
 import qualified Data.Text.IO                  as Text
-import qualified Data.Text.Lazy.IO             as TL
 import           Nix
 import           Nix.Convert
 import qualified Nix.Eval                      as Eval
@@ -124,7 +122,7 @@ main = do
     | xml opts
     = error "Rendering expression trees to XML is not yet implemented"
     | json opts
-    = liftIO $ TL.putStrLn $ A.encodeToLazyText (stripAnnotation expr)
+    = error "Rendering expression trees to JSON is not implemented"
     | verbose opts >= DebugInfo
     = liftIO $ putStr $ PS.ppShow $ stripAnnotation expr
     | cache opts

--- a/src/Nix/Expr/Types.hs
+++ b/src/Nix/Expr/Types.hs
@@ -543,7 +543,6 @@ instance ToJSON NUnaryOp
 instance ToJSON NBinaryOp
 instance ToJSON NRecordType
 instance ToJSON a => ToJSON (NExprF a)
-instance ToJSON NExpr
 
 instance (FromJSON v, FromJSON a) => FromJSON (Antiquoted v a)
 instance FromJSON a => FromJSON (NString a)
@@ -558,7 +557,6 @@ instance FromJSON NUnaryOp
 instance FromJSON NBinaryOp
 instance FromJSON NRecordType
 instance FromJSON a => FromJSON (NExprF a)
-instance FromJSON NExpr
 
 $(makeTraversals ''NExprF)
 $(makeTraversals ''Binding)

